src/Fixer/ArrayNotation/AlphabeticalArrayKeySortFixer.php

```diff
@@ -0,0 +1,312 @@
+use PhpCsFixer\FixerConfiguration\FixerOptionBuilder;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
+use PhpCsFixer\Preg;
+use PhpCsFixer\Tokenizer\CT;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Abdurrahman Uymaz <abdurrahman.uymaz@mobian.global>
+ * @author Lars Grevelink <lars.grevelink@mobian.global>
+ * @author Leander Philippo <leander.philippo@mobian.global>
+ */
+final class AlphabeticalArrayKeySortFixer extends AbstractFixer implements ConfigurableFixerInterface
+{
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        // Return a definition of the fixer, it will be used in the documentation.
+        return new FixerDefinition(
+            'Sorts keyed arrays by alphabetical order.',
+            [
+                new CodeSample(
+                    "<?php\n\$sample = array('b' => '2', 'a' => '1', 'd' => '5');\n"
+                ),
+                new CodeSample(
+                    "<?php\n\$sample = array('b' => '2', 'a' => '1', foo() => 'bar', 'd' => '5');\n",
+                    ['sort_special_key_mode' => 'special_case_on_bottom']
+                ),
+                new CodeSample(
+                    "<?php\n\$sample = array('b' => '2', 'a' => '1', foo() => 'bar', 'd' => '5');\n",
+                    ['sort_special_key_mode' => 'special_case_on_top']
+                ),
+            ],
+            null,
+            'Risky when the order of the array has an impact on the code execution.'
+        );
+    }
+
+    public function isRisky(): bool
+    {
+        return true;
+    }
+
+    public function isCandidate(Tokens $tokens): bool
+    {
+        return $tokens->isTokenKindFound(T_ARRAY) || $tokens->isTokenKindFound(CT::T_ARRAY_SQUARE_BRACE_OPEN);
+    }
+
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
+    {
+        $this->sortTokens($tokens);
+    }
+
+    protected function createConfigurationDefinition(): FixerConfigurationResolverInterface
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('sort_special_key_mode', 'Whether to sort the specials keys on the bottom `special_case_on_bottom` or top `special_case_on_top`.'))
+                ->setAllowedValues(['special_case_on_bottom', 'special_case_on_top'])
+                ->setDefault('special_case_on_bottom')
+                ->getOption(),
+        ]);
+    }
+
+    /**
+     * Sort the tokens in the correct order.
+     */
+    protected function sortTokens(Tokens $tokens): void
+    {
+        $lastProcessedIndex = null;
+        foreach ($tokens as $index => $token) {
+            if (null !== $lastProcessedIndex && $index < $lastProcessedIndex) {
+                continue;
+            }
+
+            if ($token->isGivenKind(T_ARRAY) || $token->isGivenKind(CT::T_ARRAY_SQUARE_BRACE_OPEN)) {
+                $clonedTokens = clone $tokens;
```

Comment:
What is the point of this cloning?

Reply:
My bet is that due to using recursive algo, you overcomplicate the rule and now it has to track 2 set of indexes (before fixes, after fixes).

You may do reverse loop, where you start fixing array from end of collection. Then cloning nor recursive algo should not be needed anymore

https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/5492/files#r1254383117